R/derive_param_extreme_event.R

```diff
@@ -0,0 +1,379 @@
+#'
+#' @importFrom rlang .data
+#'
+#' @details
+#'   1. The input dataset is restricted to observations fulfilling
+#'   `filter_source`.
+#'   1. For each subject (with respect to the variables specified for the
+#'   `subject_keys` parameter) either the first or last observation, depending on `mode`,
+#'    (with respect to `order`) where the event condition (`filter_source` parameter) is
+#'   fulfilled is selected.
+#'   1. For each observation in `dataset_adsl` a new observation is created. For
+#'   subjects with event `new_var` is set to `true_var`. For all other
+#'   subjects `new_var` is set to `false_var`.
+#'   For subjects with event all variables from `dataset_source` are kept. For
+#'   subjects without event all variables which are in both `dataset_adsl` and
+#'   `dataset_source` are kept.
+#'   1. The variables specified by the `set_values_to` parameter are added to
+#'   the new observations.
+#'   1. The new observations are added to input dataset.
+#'
+#' @author Stefan Bundfuss Sophie Shapcott
+#'
+#' @return The input dataset with a new parameter indicating if and when an
+#'   event occurred
+#'
+#' @family der_prm_bds_findings
+#' @keywords der_prm_bds_findings
+#'
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#' library(dplyr)
+#' library(lubridate)
+#'
+#' # Derive a new parameter for the first disease progression (PD)
+#' adsl <- tribble(
+#'   ~USUBJID, ~DTHDT,
+#'   "1",      ymd("2022-05-13"),
+#'   "2",      ymd(""),
+#'   "3",      ymd("")
+#' ) %>%
+#'   mutate(STUDYID = "XX1234")
+#'
+#' adrs <- tribble(
+#'   ~USUBJID, ~ADTC,        ~AVALC,
+#'   "1",      "2020-01-02", "PR",
+#'   "1",      "2020-02-01", "CR",
+#'   "1",      "2020-03-01", "CR",
+#'   "1",      "2020-04-01", "SD",
+#'   "2",      "2021-06-15", "SD",
+#'   "2",      "2021-07-16", "PD",
+#'   "2",      "2021-09-14", "PD"
+#' ) %>%
+#'   mutate(
+#'     STUDYID = "XX1234",
+#'     ADT = ymd(ADTC),
+#'     PARAMCD = "OVR",
+#'     PARAM = "Overall Response",
+#'     ANL01FL = "Y"
+#'   ) %>%
+#'   select(-ADTC)
+#'
+#' derive_param_extreme_event(
+#'   adrs,
+#'   dataset_adsl = adsl,
+#'   dataset_source = adrs,
+#'   filter_source = PARAMCD == "OVR" & AVALC == "PD",
+#'   order = vars(ADT),
+#'   new_var = AVALC,
+#'   true_value = "Y",
+#'   false_value = "N",
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     PARAMCD = "PD",
+#'     PARAM = "Disease Progression",
+#'     ANL01FL = "Y",
+#'     ADT = ADT
+#'   )
+#' )
+#'
+#' # derive parameter indicating death
+#' derive_param_extreme_event(
+#'   dataset = adrs,
+#'   dataset_adsl = adsl,
+#'   dataset_source = adsl,
+#'   filter_source = !is.na(DTHDT),
+#'   order = vars(DTHDT),
+#'   new_var = AVALC,
+#'   true_value = "Y",
+#'   false_value = "N",
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     PARAMCD = "DEATH",
+#'     PARAM = "Death",
+#'     ANL01FL = "Y",
+#'     ADT = DTHDT
+#'   )
+#' )
+derive_param_extreme_event <- function(dataset,
```

Comment:
The `dataset` argument should be optional (like in `derive_param_tte()`).

Reply:
@bundfussr Just a thought - in the docs currently under `dataset` it says that `PARAMCD` is expected. Now if `dataset` is optional, then surely this is no longer the case? I believe the only requirement of `dataset` is to give a dataset to bind the new rows to, no mention of `PARAMCD` at all. Therefore I have altered this description. However, please let me know if I have misunderstood.

Reply:
Ignore me, I think I can see the reasoning now from `derive_param_tte()` - it's to do with if `dataset` is not `NULL` then it checks that the `PARAMCD` provided isn't already within the dataset. Will do the same in this function.